### ISSUE TITLE
Getting current page and total number of pages

### ISSUE THREAD

**Issue by wedranb** [futurepress/epub.js#744]
Hi, I would like to show something like this (in my reader footer): `Page 5 - 323`. This is basically this: `"Page " + currentPage + " - " + totalPages`

But I didn't find a way to get `currentPage` and `totalPages` number.

For `currentPage` I tried:

```javascript
this.rendition.on('relocated', (location) => {
...
this.book.locations.locationFromCfi(location.start.cfi);
...
});
```
But I get strange numbers like (in order): 0, 1, 2, 4, 6, 8, 11...

Amnd I'm not even sure how to get `totalPages` number.

Can anyone help? I would like to find solution for both cases and add them to epub documentation.

**Comment by wedranb** [comment]
Hey, @fchasen. Thank you for this cool library you and guys created.

I was wondering if you could give me some tips for regarding my question?
I would love to help back.

Thank you!

**Comment by jlag34** [comment]
If you look at the wiki under [Updating to v0.3 from v0.2](https://github.com/futurepress/epub.js/wiki/Updating-to-v0.3-from-v0.2) you find that they have a new method:

```
generatePagination() was far too resource intensive, as it rendered every page. It has been replaced by book.locations.generate(600) which will create a CFI for every X characters in the book.
```

From my understanding depending on the argument you pass (150 being the default) it goes through the book and marks where it is every X characters. You get back a big array of pages. I don't find this very helpful though because it's not actual pages, just a rough estimate. Using this method you could skip a page if your number is too high or count a page twice if too low. It seems that when you use `rendition.currentLocation()` it has an `end and `start`. In those is the list of pages in that chapter but that can be wrong based off of your screen size. All in all, there seems to be no easy way to do it.

**Comment by fchasen** [comment]
Getting the page for a currently rendered chapter is simple enough, but rendering the entire book out in each users browser to get page numbers just wasn't practical.

Would be amazing to render out the entire book server side and save the pages as an epub page list. With [Puppeteer](https://github.com/GoogleChrome/puppeteer) it should not be too hard to setup.

**Comment by mikkelvp** [comment]
@geek1011 You can do that with numbers from `rendition.currentLocation()`

I've made an extra offscreen rendering of the book. I use that to load one section at a time and get page counts. Takes a while but it's no problem since it's all in the background after the book has loaded. Will show a spinner or similar where page counts are shown while loading them.

**Comment by askilondz** [comment]
`book.locations.generate` returns a promise containing an array of all the CFI locations. Correct me if I am wrong but if these locations are all the pages then we can just get the length of the array.

            this.book.ready.then((book) => {
                return this.book.locations.generate();
            }).then(locations => {
                console.log("Total Pages?: ", locations.length);
            });

**Comment by mikkelvp** [comment]
@askilondz You are wrong. It generates cfi ranges for every x (default 150) characters which has nothing to do with pages.

**Comment by askilondz** [comment]
@mikkelvp :-) thanks for the clarification.

**Comment by fgilio** [comment]
Hi guys! Has anyone found a practical, and hopefully accurate, solution to this?

So far the approaches I'm considering are:
	A) @fchasen solution sounds nice, but doesn't the number of pages depend also on screen and font size?
	B) Use `book.locations.generate`, though it has nothing to do with pages

**Comment by mikkelvp** [comment]
@fgilio I've described my solution earlier in this issue.
A) Yes, if you have applied styling you will need to wait for browser to render with that styling before getting page count for a section.

**Comment by askilondz** [comment]
@mikkelvp Could you show your code for how you're rendering each section for the offscreen rendering you are doing? I like this approach for getting the page count.  I'm looping through each spine item and attempting to render each item/section `item.render()` but having some issues would be curious to see your solution. Thanks!

**Comment by AK0088154** [comment]
I want to show more sections together, e.g. if the page is having an image, breaking it into sections shows part of face then neck then chest and so on, and which does not make any sense.

**Comment by sengemann** [comment]
Here's the best solution that I have come up with for getting page number and total pages based on setting the number of characters per page:

```JavaScript
// Initialize the book
let bookUri = "https://s3.amazonaws.com/moby-dick/moby-dick.epub";
let book = ePub(bookUri, {});
let rendition = book.renderTo('epubContainer', {
    flow: 'paginated',
    manager: 'continuous'
    spread: 'always'
    width: "100% - 106px",
    height: this.calculateReaderHeight()
});

// Display the book
let displayed = rendition.display(window.location.hash.substr(1) || undefined);
displayed.then(function() {
    console.log('rendition.currentLocation():', rendition.currentLocation());
});

// Generate location and pagination
book.ready.then(function() {
    const stored = localStorage.getItem(book.key() + '-locations');
    console.log('metadata:', book.package.metadata);
    if (stored) {
        return book.locations.load(stored);
    } else {
        return book.locations.generate(1024); // Generates CFI for every X characters (Characters per/page)
    }
}).then(function(location) { // This promise will take a little while to return (About 20 seconds or so for Moby Dick)
    localStorage.setItem(book.key() + '-locations', book.locations.save());
});

// When navigating to the next/previous page
rendition.on('relocated', function(locations) {
    progress = book.locations.percentageFromCfi(locations.start.cfi);
    console.log('Progress:', progress); // The % of how far along in the book you are
    console.log('Current Page:', book.locations.locationFromCfi(locations.start.cfi);
    console.log('Total Pages:', book.locations.total);
});
```

**Comment by vblazenka** [comment]
@sengemann how accurate is this? What if the user changes font size or screen size?

**Comment by sengemann** [comment]
It updates accordingly when the `relocated` event is triggered. I will have to see how to manually trigger that event for the actual values to change.

**Comment by yh54321** [comment]
> It updates accordingly when the `relocated` event is triggered. I will have to see how to manually trigger that event for the actual values to change.

As far as your understanding, if I call `rendition.currentLocation()` when on a given page, will the returned start and end CFIs be the actual start and end CFIs of the page 100% of the time? Or are these just estimates, so that if I go to one of these start CFIs, it may not bring me to the correct page...

If anyone else has an answer to this it will likewise be appreciated.

**Comment by ezequiel9** [comment]
The documentation of this library is really bad. Seems to be very  good project but I'm losing a lot of time guessing how it works :(((((

**Comment by MrXCQ** [comment]
Hi，How to get current page and total number of pages? Does anybody have a solution? PLease 🥺

**Comment by vblazenka** [comment]
@MrXCQ You can't. :) In this thread you can find possible solutions but nothing is accurate.

**Comment by sengemann** [comment]
> > It updates accordingly when the `relocated` event is triggered. I will have to see how to manually trigger that event for the actual values to change.
> 
> As far as your understanding, if I call `rendition.currentLocation()` when on a given page, will the returned start and end CFIs be the actual start and end CFIs of the page 100% of the time? Or are these just estimates, so that if I go to one of these start CFIs, it may not bring me to the correct page...
> 
> If anyone else has an answer to this it will likewise be appreciated.

The only time that the method `rendition.currentLocation()` has not been returning the correct location was immediately after adjusting the font-size via the `themes()` method. I have a bug created for that here: https://github.com/futurepress/epub.js/issues/982

**Comment by hesampour** [comment]
@sengemann would you please give us the calculateReaderHeight() function?

**Comment by sengemann** [comment]
@hesampour, it's pretty simple once you know what your variables are:

``` JavaScript
calculateReaderHeight() {
  let h = Math.max(document.documentElement.clientHeight, window.innerHeight || 0);
  // On the next line you need to factor in the height of any of your UI elements.
  // You could just take the  values if they are static or use jQuery to get the heights dynamically (that would be the preferred method)
  h = h - 50 - 68 - 44; // 50: Header Height, 68: Footer Height, 44: 22px * 2 top/bottom margin
  return h;
},
```

**Comment by jiangxiaopeng** [comment]
> The documentation of this library is really bad. Seems to be very good project but I'm losing a lot of time guessing how it works :(((((

thats right

**Comment by hrkazemi** [comment]
does anyone have a solution for **pagination**?

**Comment by sbrighiu** [comment]
pagingEnabled={true} ?

**Comment by sbrighiu** [comment]
I'm messing around with the react native implementation, and adjusting it to my needs here https://github.com/sbrighiu/epubjs-rn.

Would be really nice to be able to get number of pages and current page, but as my content varies from device size to device size and from epub to epub, I don't see a reliable way of doing it :(

**Comment by sbrighiu** [comment]
What about the completion? Like a percentage? 61%

**Comment by sengemann** [comment]
@sbrighiu and @MrXCQ This is the best solution that I have found to get the percentage, current page, page range and total pages: https://github.com/futurepress/epub.js/issues/744#issuecomment-492300092

There are a few requirements for this to work:

- You need to set your height and width of the area to render the book (I have the set by some function returns in my example code in my comment)
- You need to set a "page size" with `book.locations.generate(X)` where `X` is the number of characters you want to have on a page
- The entire book needs to be loaded (not one chapter at a time)

**Comment by hrkazemi** [comment]
> What about the completion? Like a percentage? 61%
@sbrighiu 
I did use the [mupdf](https://mupdf.com/docs/manual-mutool-run.html) library instead. it's work for both epub & pdf (and great if you are doing native but has poor document).

**Comment by JoaoVictorSarzi** [comment]
@fchasen help me please... I have to different .epub, but crafted the same way... i dont know why one of then dont generate the locations, get the spine and everything but location returns -1 only. The promise dont complete and dont get in on .then().
Promisse sshow this on every part: [Exception: TypeError: 'caller', 'callee', and 'arguments' properties may not be accessed on strict mode functions or the arguments objects for calls to them at Function.o (<anonymous>:1:83)]

**Comment by OzkanAbdullahoglu** [comment]
seems like this one is still open.  While I was trying to figure things out, I just found a solution that looks like enough for my current problem. 
In the Rendition object, we have location as a static member. In location, there are start and end objects so in these objects we have an index ( which I assume right now as a section)  and in displayed there is a page and total numbers which belong to this index. 

You can check the [docs](http://epubjs.org/documentation/0.3/#renditionlocation) again for the detailed structure. 

Also here below I inserted a screenshot for a better understanding. 

Meanwhile, If anyone got a better idea please let me know. 

<img width="1430" alt="Screen Shot 2021-05-18 at 03 10 50" src="https://user-images.githubusercontent.com/34690158/118571889-ce078780-b787-11eb-81cc-42a50b7f4552.png">

**Comment by RatulSikder97** [comment]
> seems like this one is still open. While I was trying to figure things out, I just found a solution that looks like enough for my current problem. In the Rendition object, we have location as a static member. In location, there are start and end objects so in these objects we have an index ( which I assume right now as a section) and in displayed there is a page and total numbers which belong to this index.
> 
> You can check the [docs](http://epubjs.org/documentation/0.3/#renditionlocation) again for the detailed structure.
> 
> Also here below I inserted a screenshot for a better understanding.
> 
> Meanwhile, If anyone got a better idea please let me know.
> 
> <img alt="Screen Shot 2021-05-18 at 03 10 50" width="1430" src="https://user-images.githubusercontent.com/34690158/118571889-ce078780-b787-11eb-81cc-42a50b7f4552.png">

Here, you only get the Total page and current page for the current chapter, not the overall. 😥

**Comment by imron199** [comment]
this is not working

**Comment by elliotdickison** [comment]
Some thoughts after beating my head against this problem for a few hours:

**Page numbers are a lost cause**

IMO "page number" is not a particularly helpful concept when it comes to ebooks. An "accurate" page number that increments or decrements by exactly 1 every time the user moves backwards or forwards by 1 page must be calculated on the client. There is no way around this - this type of page number is dependent on the exact rendering of the book which is dependent on the client's viewport, the client's interpretation of fonts, the client's interpretation of CSS, etc.

Even if you were willing to deal with this limitation and churn out page numbers on the client, the results are not that helpful to the user. You are showing them an arbitrary page number that won't match up with page numbers generated on another other device or even on the same device if the view is rotated resized. Imagine reading a book, noting that you were on page 3 of 450, rotating your device, and then discovering that you are now on page 2 of 300. This is confusing (and unavoidable).

**A better metric**

I think total % progress in the book is a much better metric. This can be calculated with locations as demonstrated in some of the responses above and shouldn't change appreciably if the user switches devices, rotates their screen, resizes their browser, etc. For a little more context you could combine this with "X pages left in chapter" which doesn't require any extra calculations (although this number would change if the view is resized/rotated/etc.).

**Generating locations server side**

@fchasen mentioned using Puppeteer to generate locations on the server. I'm sure this would work, but Puppeteer tends to be overkill. I did a little testing and with some patience it's possible to use [jsdom](https://www.npmjs.com/package/jsdom) and [xhr2](https://www.npmjs.com/package/xhr2) to get epubjs working in Node itself (at least enough to where you can generate locations and extract the book's table of contents).

Hope these ramblings are somewhat helpful.

**Comment by evanreichard** [comment]
I've been battling with this for a bit so I'm still trying to find better ways to do this as quickly and accurately as possible. The best I've come up with:

Since we have access to:
- Total Locations (after `book.locations.generate(1024)`)
- Current Section Locations
- Current Section Pages

We can estimate the current and total pages:
```
TotalLocations / TotalPages = SectionLocations / SectionPages
Total Pages = (TotalLocations * SectionPages) / SectionLocations;
```

The implementation:
```
async function estimatePages() {
  if (book.locations.total == 0) await book.locations.generate(1024);

  let currentLocation = rendition.currentLocation();
  let sectionIndex = currentLocation.start.index;
  let sectionPages = currentLocation.start.displayed.total;
  let sectionBaseCFI = book.spine.get(sectionIndex).cfiBase;

  let sectionStartLocation = book.locations._locations.findIndex((item) =>
    item.startsWith("epubcfi(" + sectionBaseCFI)
  );
  let sectionLocations = book.locations._locations.filter((item) =>
    item.startsWith("epubcfi(" + sectionBaseCFI)
  ).length;

  let totalLocations = book.locations.total;
  let estPages = (totalLocations * sectionPages) / sectionLocations;
  let estSectionStartPage =
    estPages * (sectionStartLocation / totalLocations);
  let estCurrentPage =
    estSectionStartPage + currentLocation.start.displayed.page;
  return {
    currentPage: Math.round(estCurrentPage),
    totalPages: Math.round(estPages),
  };
}
```

Benefits / Notes:
- Super fast to recalculate if the font / zoom changes
- The more sections you have (`1024` in above example), the more accurate it should be.
- The longer the currently rendered section is, the more accurate it should be.

It's not perfect, but it's the best I could come up with.

**Comment by evanreichard** [comment]
@vedmant Yeah there's some wonkyness when changing between chapters. I ended up abandoning total page count and reverting to total percentage, and the current / total page counts for the current chapter only.

**Comment by chrisovato** [comment]
Any updates on this case?

**Comment by zgndt** [comment]
Hello everyone. I struggled a lot to make an epub reader for all platforms. The view is depends on device screen or container you want to render. The character counting methods are all wrong. First of all you need to know there are 2 kind of epub format in general. One is fake epub which converts from pdf with some online free softwares. They give absolute positions for each words and scale the page with around 200-500px font-sizes so you can not do font-size/font-family or any font related changing. So keep in mind that you must disable settings for these ones.  Page calculation goes right but rest of epub features will blow up your coding just because of they are simply fake! 

So at start please spare time to detect that!

After all of these i am using an algoritm with an offscreen loader (an asyc task) 
declaring an object array to keep each page start cfi locations and a simply page number
you can use next async function and its "then" method to create a loop
if user try to open page numbers screen. I am simply showing a loader and "Ebook is re-converting for your device screen, please wait" it is not taking too long. This is just for frenzy users

After that when you change font style or padding/margin which may effect total page count. (not for font or bg color)
You must re-call that algoritm and you may use a splash that explains "Re-Converting" to user 

You can wait till all page count calculated while in splash (that period is depends on the book real page size)
or just fade-out splash after first and next 10 pages are calculated to not disturb user long. Async calculation will not break user actions and they can start reading for sure.

If you want to record the user last used page's START location to give them a feature like "Continue reading" when they re-open your app later, simple record start CFI of screen but be careful they may open that book from another device!!!

So you must calculate device screen oriented page number by using old recorded cfi (last location)
So you must write a function that converts cfi to real number and it is simple too. If you look careful the cfi string logic is just device pages into paragraphs and paragraphs into parts with using splitters ('/' for sections and pages, ':' for pharagraphs or something like that). Just split the string, take the numbers, control the numbers in a loop and currentPage++ until you pass the entire location then stop and display(that_location) & "Remove your splash"

Thats all

You don't need a copy code. You must do your own logic depends on your scenario and device types and user types. Epub.js gives enough information for you to do all of these.

Thank you & hope these helps

**Comment by RomanSoviak** [comment]
Any updates?
@zgndt any pieces of code would be appreciated
